inc/saiobject.h

```diff
@@ -145,6 +166,34 @@ sai_status_t sai_bulk_get_attribute(
         _Inout_ sai_attribute_t **attr_list,
         _Inout_ sai_status_t *object_statuses);
 
+/**
+ * @brief Query attribute capability
+ *
+ * @param[in] object_type SAI object type
+ * @param[in] attr_id SAI attribute ID
+ * @param[out] attr_capability Capability per operation
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+sai_status_t sai_query_attribute_capability(
+        _In_ sai_object_type_t object_type,
+        _In_ sai_attr_id_t attr_id,
+        _Out_ sai_attr_capability_t *attr_capability);
+
+/**
+ * @brief Query an enum attribute (enum or enum list) list of implemented enum values
+ *
+ * @param[in] object_type SAI object type
+ * @param[in] attr_id SAI attribute ID
+ * @param[inout] enum_values_capability List of implemented enum values
+ *
+ * @return #SAI_STATUS_SUCCESS on success, #SAI_STATUS_BUFFER_OVERFLOW if list size insufficient, failure status code on error
+ */
+sai_status_t sai_query_attribute_enum_values_capability(
+        _In_ sai_object_type_t object_type,
+        _In_ sai_attr_id_t attr_id,
+        _Inout_ sai_s32_list_t *enum_values_capability);
```

Comment:
Im not following description here, can you give example of such api ?

From current description im assumming that if attribute is enum or enum list, this api will return only supported enum values for that specific attribute id ?

Reply:
Let's look as an example on

/**
 * @brief Priority flow control mode
 */
typedef enum _sai_port_priority_flow_control_mode_t
{
    /** Same value for RX/TX */
    SAI_PORT_PRIORITY_FLOW_CONTROL_MODE_COMBINED,

    /** Separate values for RX/TX */
    SAI_PORT_PRIORITY_FLOW_CONTROL_MODE_SEPARATE,

} sai_port_priority_flow_control_mode_t;

    /**
     * @brief Combined or separate Bit vectors for port PFC RX/TX
     *
     * @type sai_port_priority_flow_control_mode_t
     * @flags CREATE_AND_SET
     * @default SAI_PORT_PRIORITY_FLOW_CONTROL_MODE_COMBINED
     */
    SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL_MODE,

Now let's say your SAI implementation supports only combined mode, and not separated mode

Application will :
sai_s32_list_t attribs
attribs.count = 2
attribs.list = malloc 2 items
call the query enum values api

SAI implementation in this example, will fill
count = 1
list[0] = combined mode

Reply:
great thanks for explaining

Reply:
Hi @itaibaz

  I came across this explanation and would to understand how can we achieve a capability check where there are multiple values for an attribute, and the SAI capability check will return back "True" only if it supports all of those attributes together.

  The scenario is like this --> for a particular ASIC family, the SAI does not support both v4 (SRC_IP, DST_IP) and V6(SRC_IPV6, DST_IPV6) attributes in one field group for user ACL tables. The key size is exceeding the HW max-key.
 
So can we have a capability query which checks if the combination of "SAI_ACL_ACTION_TYPE_SET_SRC_IP, SAI_ACL_ACTION_TYPE_SET_DST_IP, SAI_ACL_ACTION_TYPE_SET_SRC_IPV6, SAI_ACL_ACTION_TYPE_SET_DST _IPV6" for the attribute SAI_ACL_TABLE_ATTR_ACL_ACTION_TYPE_LIST is supported OR not. 

Currently I understand that we have the capability check individually for a capability of an attribute, or as in your eg: above the capability query can return back what all capabilities SAI supports for an attribute. 

But how do we do this capability check for a combination of capabilities that could be used together. Pls share your thoughts - I can share more details too.

 -thanks

Reply:
Hi,
The current method is not intended to query relationship between values and more complicated limitations
In the example you mentioned, the query should return all values, since they are individually supported. And the additional limitations can't be expressed
In addition, the query capability is intended to query specific vendor implementation, that might be a subset of the full SAI API. But the case you describe is generic to all vendors, you can't set both v4 and v6 fields, as a packet is either v4 or v6. So this could theoretically be described in the SAI API itself IMO

Reply:
Thanks @itaibaz